<a href="https://colab.research.google.com/github/AntonySSk/Machine_Learning/blob/dev/%22%D0%9B%D0%912_2_%D0%A8%D0%BA%D0%BE%D0%BB%D0%B0_3_9_%D0%9C%D0%9D_%D0%97%D0%B0%D0%B2%D0%B4%D0%B0%D0%BD%D0%BD%D1%8F_2_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Лабораторна робота 2.2. Школа, 3-9

---



In [21]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os

1. Вивести перші 5 рядків датасета

In [22]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("mahmoudshogaa/titanic-dataset")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'titanic-dataset' dataset.
Path to dataset files: /kaggle/input/titanic-dataset


In [23]:
df = pd.read_csv(os.path.join(path, "titanic.csv"))
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


2.	Визначити розмір датасета.

In [24]:
df.shape

(891, 12)

 3.	Визначити тип даних.

In [25]:
df.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


4.	Визначити наявність пропущених значень. При наявності, замінити пропущені значення на середнє значення.

In [26]:
# Перевіряємо пропущені значення
print(df.isnull().sum())

# Заповнюємо пропуски в числових стовпцях їхнім середнім значенням
df = df.fillna(df.mean(numeric_only=True))

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


5. Ще раз перевірити наявність пропущених значень.

In [27]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


6.	Сформувати датасет з обраними стовпцями ['Survived','Pclass','Sex','Age', 'Fare']

In [28]:
selected_df = df[['Survived', 'Pclass', 'Sex', 'Age', 'Fare']]
selected_df.head()

,Survived,Pclass,Sex,Age,Fare
0,0,3,male,22.0,7.2500
1,1,1,female,38.0,71.2833
2,1,3,female,26.0,7.9250
3,1,1,female,35.0,53.1000
4,0,3,male,35.0,8.0500


7.	Замінити біранарні ознаки (Стать) на 0 і 1 (з перевіркою унікальних значень даного стовпчика).

In [29]:
df['Sex'].value_counts()


,count
Sex,
male,577
female,314


In [30]:
# Convert categorical data to numerical data
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

print(f'Data types of the dataset:\n{df.dtypes}')

Data types of the dataset:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex              int64
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


8. Вивести описову статистику датасету describe()

In [31]:
df.describe()

,PassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,0.352413,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,0.477990,13.002015,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,0.000000,22.000000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,0.000000,29.699118,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,1.000000,35.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,1.000000,80.000000,8.000000,6.000000,512.329200


9. Ще раз перевірити кількість пропущених даних (впевнитись, що їх немає).

In [32]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


10. Вивести 5 перших рядків датасету.

In [33]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,NaN,S


11. Вивести 5 останніх рядків датасету.

In [34]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",0,27.000000,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",1,19.000000,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",1,29.699118,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",0,26.000000,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",0,32.000000,0,0,370376,7.75,NaN,Q


12. Аналіз виживання залежно від статі: Обчисліть відсоток виживання для кожної статі. Чи була різниця у виживанні між чоловіками та жінками?

In [35]:
survival_by_sex = df.groupby('Sex')['Survived'].mean() * 100
print(survival_by_sex)

Sex
0    18.890815
1    74.203822
Name: Survived, dtype: float64


Так, була суттєва різниця. Відсоток виживання серед жінок (близько 74%) є значно вищим, ніж серед чоловіків (близько 19%).

13. Обчисліть відсоток виживання для кожного класу (Pclass). Який клас мав найвищий рівень виживання (дати відповідь)?

In [36]:
survival_by_pclass = df.groupby('Pclass')['Survived'].mean() * 100
print(survival_by_pclass)

Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64


Найвищий рівень виживання мав 1-й клас. Відсоток виживання становить близько 63% для першого класу, 47% для другого класу та лише 24% для третього класу.

14.	Визначте середній вік тих, хто вижив, і тих, хто не вижив. Чи впливає вік на виживання (дати відповідь)?

In [37]:
mean_age_by_survived = df.groupby('Survived')['Age'].mean()
print(mean_age_by_survived)

Survived
0    30.415100
1    28.549778
Name: Age, dtype: float64


Так, вік впливає на виживання,середня різниця між вижившими та загиблими  невелика (середній вік тих, хто вижив - близько 28.55 років, а тих, хто загинув - близько 30.41 років). Головна причина в тому, що дітей евакуювали першочергово, що суттєво підвищило відсоток виживання серед наймолодших пасажирів.

15.	Розподіліть пасажирів на групи за рівнями тарифів (Fare) і обчисліть рівень виживання для кожної групи. Як тариф впливав на шанси виживання (дати відповідь)?

In [38]:
# Розподіляємо вартість квитка (Fare) на 4 квантильні групи (Низька, Середня, Вище середнього, Висока)
df['FareGroup'] = pd.qcut(df['Fare'], q=4, labels=['Низький', 'Середній', 'Вище середнього', 'Високий'])

# Обчислюємо відсоток виживання для кожної групи
survival_by_fare = df.groupby('FareGroup', observed=False)['Survived'].mean() * 100
print(survival_by_fare)

FareGroup
Низький            19.730942
Середній           30.357143
Вище середнього    45.495495
Високий            58.108108
Name: Survived, dtype: float64


Так, тариф суттєво впливав на шанси виживання: чим вищою була ціна квитка (тариф Fare), тим вищим був відсоток виживання. Пасажири з категорії з найвищими тарифами мали значно більше шансів врятуватися (понад 50-60%), оскільки зазвичай це були пасажири 1-го класу, каюти яких розташовувалися вище та ближче до шлюпочної палуби.

16. Аналіз класу та тарифу: Визначте середній тариф (Fare) для кожного класу (Pclass). Чи існує значна різниця у тарифах між класами (дати відповідь)?

In [39]:
mean_fare_by_pclass = df.groupby('Pclass')['Fare'].mean()
print(mean_fare_by_pclass)

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64


Так, існує дуже значна різниця у тарифах між класами. Середня вартість квитка для 1-го класу становить близько 84.15, для 2-го класу - близько 20.66, а для 3-го класу - лише близько 13.68. Тариф для першого класу суттєво перевищує вартість квитків другого та третього класів.

17. Обчисліть середній вік пасажирів для кожного класу (Pclass). Як вік пасажирів розподіляється за класами (дати відповідь)?

In [41]:
mean_age_by_pclass = df.groupby('Pclass')['Age'].mean()
print(mean_age_by_pclass)

Pclass
1    37.048118
2    29.866958
3    26.403259
Name: Age, dtype: float64


Вік пасажирів чітко зменшується зі зниженням класу. Середній вік пасажирів 1-го класу є найвищим і становить близько 37 років, у 2-му класі - близько 30 років, а в 3-му класі пасажири наймолодші - середній вік становить близько 26 років.